_posts/2025-04-25-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,399 @@
+
+We are ready to test our secure MCP server in the dev mode.
+
+== Step 2 - Access MCP server in the dev mode
+
+=== Start MCP server in the dev mode
+
+[source,shell]
+----
+mvn quarkus:dev
+----
+
+The configuration properties that we set in the <<initial-configuration>> section above are sufficient to start the application in the dev mode.
+
+The OIDC configuration is provided in the dev mode automatically by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak]. It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started with OIDC immediately. You can also register a custom Keycloak realm to work with the existing realm, client and user registrations.
+
+No problems if you do not work with Keycloak, see the <<mcp-server-devui>> section for more details.
+
+[[oidc_devui]]
+=== Use OIDC Dev UI to login and copy access token
+
+Go to http://localhost:8080/q/dev[Dev UI], find the OpenId Connect card:
+
+image::oidc_devui.png[OIDC in DevUI,align="center"]
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+[NOTE]
+====
+You can login to other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] from OIDC DevUI as well. The only requirement is to update your application registration to allow callbacks to DevUI. For example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0 from Dev UI].
+====
+
+After logging in with `Keycloak` as `alice`, copy the acquired access token using a provided copy button:
+
+image::login_and_copy_access_token.png[Login and copy access token,align="center"]
+
+[[mcp-server-devui]]
+=== Use Quarkus MCP Server Dev UI to access MCP server
+
+Make sure to login and copy the access token as explained in the <<oidc-devui>> section above.
+
+Go to http://localhost:8080/q/dev[Dev UI], find the MCP Server card:
+
+image::mcp_server_devui.png[MCP Server in DevUI,align="center"]
+
+Select its `Tools` option and choose to `Call` the `user-name-provider` tool:
+
+image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]
+
+Paste the copied Keycloak access token into the Tool's `Bearer token` field, and request a new MCP SSE session:
+
+image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]
+
+Make a tool call and get a response which contains the `alice` user name:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+Now, please stop the server: we are going to prepare it to run in the prod mode next.
+
+== Step 3 - Access MCP server in the prod mode
+
```

Comment:
Maybe add something like:

"Before it was all in dev mode - using Quarkus devservices to try things out. Now lets move to prod mode. If you have an keycloak instance running you can use that (see link); but to illustrate how oauth works with more than just keycloak and something many of you have direct access we will show prod mode using GitHub OpenID."

?